autopilot/scanner_test.go

```diff
@@ -0,0 +1,150 @@
+package autopilot
+
+import (
+	"errors"
+	"net"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"gitlab.com/NebulousLabs/fastrand"
+	"go.sia.tech/renterd/bus"
+	"go.sia.tech/renterd/hostdb"
+	"go.sia.tech/renterd/internal/consensus"
+	"go.sia.tech/renterd/worker"
+)
+
+type mockBus struct {
+	hosts []hostdb.Host
+
+	mu   sync.Mutex
+	itxs []hostdb.Interaction
+}
+
+func (b *mockBus) AllHosts() ([]hostdb.Host, error) { return b.hosts, nil }
+func (b *mockBus) ConsensusState() (bus.ConsensusState, error) {
+	return bus.ConsensusState{BlockHeight: 0, Synced: true}, nil
+}
+func (b *mockBus) RecordHostInteraction(_ consensus.PublicKey, itx hostdb.Interaction) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.itxs = append(b.itxs, itx)
+	return nil
+}
+
+func (b *mockBus) counts() (success int, failed int) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	for _, itx := range b.itxs {
+		if itx.Success {
+			success++
+		} else {
+			failed++
+		}
+	}
+	return
+}
+
+type mockWorker struct{}
+
+func (w *mockWorker) RHPScan(_ consensus.PublicKey, hostIP string, _ time.Duration) (r worker.RHPScanResponse, e error) {
+	if strings.HasSuffix(hostIP, "fail") {
+		e = errors.New("fail")
+	}
+	if strings.HasSuffix(hostIP, "sleep") {
+		time.Sleep(time.Second)
+	}
+	return
+}
+
+func TestScanner(t *testing.T) {
+	h := testHosts(100)
+	b := &mockBus{hosts: h}
+	w := &mockWorker{}
+	s := &scanner{
+		bus:             b,
+		worker:          w,
+		tracker:         newTracker(trackerMinDataPoints, trackerNumDataPoints, trackerTimeoutPercentile),
+		stopChan:        make(chan struct{}),
+		scanMinInterval: time.Second,
+	}
+	s.pool = &scanPool{numThreads: 3, s: s}
+
+	h[0].Announcements[0].NetAddress += "fail"
+	h[1].Announcements[0].NetAddress += "fail"
+	h[2].Announcements[0].NetAddress += "fail"
+
+	// start scan and assert consecutive calls are no-ops
+	if !s.tryPerformHostScan() || s.tryPerformHostScan() {
+		t.Fatal("unexpected")
+	}
+
+	// wait until the scan is done
+	var c int
+	for s.isScanning() && c < 100 {
+		time.Sleep(time.Millisecond * 10)
+		c++
+	}
```

Comment:
Ideally we would never call `time.Sleep` in a test. (*Maybe* in high level integration tests, but certainly not when using mocks.) `time.Sleep` is the number one cause of NDFs.

Reply:
I rewrote the scanner tests to avoid `time.Sleep` except for `case <-time.After(x)` to ensure tests don't hang